### PR TITLE
Removing suspension prioritization scheme for GC threads vs. non-GC

### PR DESCRIPTION
**Pull request by VSadov**
Follow up for: https://github.com/dotnet/runtime/issues/43998

There is a mechanism that tries to give priority to GC suspensions vs. other kinds. It does not look like it is worth the complexity. 
It could have been more important in the past when suspension could often take 10+ msec. (vs 0.5 - 0.01 msec now)

I do not think the problem which this mechanism tries to mitigate is real these days.

The mechanism could never possibly work with server GC anyways, which heavy threaded workloads are likely to use.
The whole scheme relies on setting `m_pThreadAttemptingSuspendForGC` to the current thread when GC thread tries to suspend EE. Yet current thread is `NULL` for GC threads in server mode, thus throttling would not activate.

The main idea is that a non-GC suspension would check `m_pThreadAttemptingSuspendForGC` after acquiring ThreadStoreLock (TSL), see that workstation GC is trying to suspend, decide to not do its own suspension and release the TSL. In such case the non-GC thread waits on `hAbortEvt`, which GC thread releases after successfully acquiring TSL.

I think this part works as intended, but effective only within a short burst of suspension requests. Suspensions that are not dense enough to contend for TSL would happen in the order they arrive.
Sustained TSL contention via suspension attempts seems to be hard to arrange in real scenarios or even intentionally. 

NOTE: The threads that just acquire TSL (not for suspension) would not check `m_pThreadAttemptingSuspendForGC` and yield the lock after acquiring, so they are unaffected by this either way.

### CODE REVIEW

**Comment by VSadov** [comment]
CC: @davidwrighton @jkotas

**Comment by mangod9** [comment]
Assume this change doesnt have a perf benefit (or regression), but simplifies the suspension logic?

**Comment by VSadov** [comment]
@mangod9 - yes, presence of this code does not seem to make any difference performance-wise. 
The condition it tries to handle was rare in the past and extremely rare now. Potential damage is very small too. Evidently, we did not have it working for server GC and never noticed...

It brings a good deal of complexity though - additional event to block on, with a cache, additional thread states when suspending.

**Comment by davidwrighton** [comment]
@vsadov What testing around debugger/gc suspesion collisions did you run? In general, I think the change looks good, but it appears to me that a meaningful amount of this logic is tied to how debugger and gc suspensions can conflict, and I don't fully understand the impacts there.

**Comment by VSadov** [comment]
The interaction with debugger is a bit hard to follow indeed:
- debugger thread would not wait on the throttling event (`hAbortEvt`) since that part treats `SUSPEND_FOR_DEBUGGER_SWEEP` in the same bucket as GC.
- another part does not treat debugger sweep the same as GC though - debugger thread will yield its attempt at suspension if it sees that GC wants to suspend. 
Since debugger thread would not block on the `hAbortEvt` there is a chance it will try taking TSL and suspending again. Sometimes locks favor the recent holder, so the whole deal might take a few iterations before GC wins.
- once GC suspends the EE it may notice that some threads are in unsafe (unwalkable) state due to breakpoints and will have to resume EE to let debugger do its thing. That may ultimately give priority to the debugger.

With server GC we do only the last part, which is required for correctness.

I will run debugger tests with this change to see if there is any impact.

**Comment by VSadov** [comment]
I've  ran debugger tests:

- ran baseline with whatever runtime gets pulled in by default, about 10 times and there were no failures!!!
my machine often had baseline failures in the past, so I always do some baseline runs first.
Whatever was done to the test stability, I like it! (@tommcdon  @hoyosjs 👍 )

- ran all the tests with my change 30 times - no failures!!
Also the end-to-end time was shorter by about 30%  -  160sec  ---> 120sec

- ran the tests a few times with a local build from a commit right before my changes, same results as with my changes.

So, something makes the perf to improve a lot, compared to the default baseline, but it is not in this change. 
This change seems perf-neutral. No failures too.

**Comment by mangod9** [comment]
@tommcdon can you please get someone from debugger side to approve as well. Looks good from VM side.

**Comment by tommcdon** [comment]
Adding @hoyosjs to review the PR

**Comment by davidwrighton** [comment]
@VSadov I can't find any uses of SUSPEND_FOR_DEBUGGER_SWEEP. How does that correlate with your analysis. Also, should we even have SUSPEND_FOR_APPDOMAIN_SHUTDOWN as an enum value if it isn't set anywhere?

**Comment by VSadov** [comment]
The comment says that `SUSPEND_FOR_DEBUGGER_SWEEP` should be used in `SysSweepThreadsForDebug`, but it is indeed not the case. The actual use might have rotted away long time ago. Same with `SUSPEND_FOR_APPDOMAIN_SHUTDOWN` and maybe some others. 
Distinguishing the reasons for suspension would be similar to tagging locks with a purpose why they are held. I think the primary use case is tracing and maybe asserts, not the functional differences, so enum values may get quietly obsolete when scenario disappears.

My analysis was that it is not clear what the code achieves at the end and what actually ends up with higher priority - GC or debugger. It tries to special-case `SUSPEND_FOR_DEBUGGER_SWEEP`, but unclear to what end, even if that was in use.

I think it does not matter, except for the case where GC must yield to debugger for correctness reasons. 
Otherwise collisions are very rare and suspensions happen in the order of requesting them and thus we do not see any sensitivity to having or not having this mechanism.

**Comment by VSadov** [comment]
I looked at the uses of `SuspendEE` and `SUSPEND_FOR_APPDOMAIN_SHUTDOWN` is actually used in unloadable assemblies stuff. The code says that it is just a value that was picked.
```
// SuspendEE cares about the reason flag only when invoked for a GC
// Other values are typically ignored. If using SUSPEND_FOR_APPDOMAIN_SHUTDOWN
// is inappropriate, we can introduce a new flag or hijack an unused one.
```
I think this could just use `SUSPEND_OTHER`.

`SUSPEND_OTHER` is the most popular. There are also uses of `SUSPEND_FOR_PROFILER`, `SUSPEND_FOR_GC` and `SUSPEND_FOR_GC_PREP`. 
There is some inconsistency in use though. For example updating write barriers on ARM uses `SUSPEND_OTHER` and `SUSPEND_FOR_GC_PREP` on x64.

Some enum values are used for purposes other than suspension - like to tag TSL when holding it.

I think it may be worth to clean this up, but as a different change. We would need to rationalize the purpose a bit - whether we want more enum values, or less - what are the kinds of uses we really care to track. 
Some values may be known to diagnostics/profiler tools so that should also be considered.

**Review by hoyosjs — APPROVED** [review]
Not my area of expertise. The changes make sense and I don't see anything that will break the debugger. Will monitor the tests to see if there's any weird behavior and let you know.

**Comment by VSadov** [comment]
@hoyosjs Thanks!!

**Review by mangod9 — APPROVED** [review]
Thanks for preemptively making readability better.

**Comment by VSadov** [comment]
Thanks!!!
